packages/web_ui/src/components/ModPackViewPage.tsx

```diff
@@ -239,19 +253,260 @@ function SearchModsTable(props: SearchModsTableProps) {
+		};
+		// Re-fetch when modal opens or factorio version changes
+	}, [open, factorioVersion, control]);
+
+	// Function to handle download from Mod Portal to Controller
+	function handleControllerDownload() {
+		if (!factorioVersion || !mods) {
+			return;
+		}
+
+		// Enable any required builtins
+		for (const name of builtins) {
+			props.onChange({
+				type: "mods.set",
+				name: name,
+				value: { ...props.modPack.mods.get(name)!, enabled: true },
+			});
+		}
+
+		// Disable any incompatible mods present
+		for (const name of incompatible) {
+			const modRecord = props.modPack.mods.get(name);
+			if (modRecord) {
+				props.onChange({
+					type: "mods.set",
+					name: name,
+					value: { ...modRecord, enabled: false },
+				});
+			}
+		}
+
+		// Add all dependencies to the mod pack
+		for (const mod of allMods) {
+			props.onChange({
+				type: "mods.set",
+				name: mod.name,
+				value: {
+					name: mod.name,
+					version: mod.version,
+					sha1: mod.sha1,
+					enabled: true,
+				},
+			});
+		}
+
+		// Download the missing dependencies
+		control.send(
+			new lib.ModPortalDownloadRequest(
+				mods.map(mod => ({
+					name: mod.name, version: new lib.ModVersionEquality("=", mod.version),
+				})),
+				factorioVersion,
+			)
+		).then(() => {
+			notification.success({
+				message: "Download complete",
+				description: `${mods.length} mods have been downloaded to the controller.`,
+			});
+		}).catch(
+			notifyErrorHandler("Error starting mod download")
+		);
+	};
+
+	return <>
+		{props.checkForUpdates && <Button
+			disabled={props.disabled}
+			icon={<CloudSyncOutlined />}
+			onClick={() => { setOpen(true); }}
+		>Check for Updates</Button> || <Button
+			disabled={props.disabled}
+			icon={<CloudDownloadOutlined />}
+			onClick={() => { setOpen(true); }}
+		>Download Dependencies</Button>}
+		<Modal
+			title={props.checkForUpdates ? "Download Updates" : "Download Dependencies"}
+			open={open}
+			onCancel={() => { setOpen(false); }}
+			width={1000}
+			footer={[
+				<Button key="close" onClick={() => { setOpen(false); }}>
+					Close
+				</Button>,
+				<Button
+					key="start"
+					type="primary"
+					disabled={loading || mods.length === 0}
+					onClick={() => { setOpen(false); handleControllerDownload(); }}
+				>Start Download</Button>,
+			]}
+		>
+			{/* Display error message if fetching failed */}
+			{error && <Alert
+				message="Error Fetching Mods"
+				description={error.message}
+				type="error"
+				showIcon
+				style={{ marginBottom: 16 }}
+			/>}
+
+			{/* Display warning message if dependencies could not be found */}
+			{!loading && missing.length > 0 && <Alert
+				message="Failed to Find Dependencies:"
+				description={<ul>
+					{missing.map(name => <li>{name}</li>)}
+				</ul>}
+				type="warning"
+				showIcon
+				style={{ marginBottom: 16 }}
+			/>}
+
+			{/* Display warning message if dependencies are incompatible */}
+			{!loading && missing.length > 0 && <Alert
```

Comment:
this should be incompatible.length, no?

Reply:
Yes, this was a copy paste error from the code above, will fix before merge.